wgpu-hal/src/gles/web.rs

```diff
@@ -0,0 +1,296 @@
+use glow::HasContext;
+use parking_lot::Mutex;
+use wasm_bindgen::JsCast;
+
+use super::TextureFormatDesc;
+
+/// A wrapper around a [`glow::Context`] to provide a fake `lock()` api that makes it compatible
+/// with the `AdapterContext` API fromt the EGL implementation.
+pub struct AdapterContext {
+    pub glow_context: glow::Context,
+}
+
+impl AdapterContext {
+    /// Obtain a lock to the EGL context and get handle to the [`glow::Context`] that can be used to
+    /// do rendering.
+    #[track_caller]
+    pub fn lock(&self) -> &glow::Context {
+        &self.glow_context
+    }
+}
+
+#[derive(Debug)]
+pub struct Instance {
+    canvas: Mutex<Option<web_sys::HtmlCanvasElement>>,
+}
+
+// SAFE: WASM doesn't have threads
+unsafe impl Sync for Instance {}
+unsafe impl Send for Instance {}
+
+impl crate::Instance<super::Api> for Instance {
+    unsafe fn init(_desc: &crate::InstanceDescriptor) -> Result<Self, crate::InstanceError> {
+        Ok(Instance {
+            canvas: Mutex::new(None),
+        })
+    }
+
+    unsafe fn enumerate_adapters(&self) -> Vec<crate::ExposedAdapter<super::Api>> {
+        let canvas_guard = self.canvas.lock();
+        let gl = match *canvas_guard {
+            Some(ref canvas) => {
+                let window = web_sys::window().expect("Could not get browser window");
+
+                if canvas.get_attribute("width").is_none() {
```

Comment:
this seems like it should be done at `Surface::configure()` time rather than here

Reply:
Oh yeah, that canvas is actually purely for enumerating adapters and doesn't actually even need a size, so we can just take that whole section out. Also, it appears that the canvas's size is already properly set by winit or something so I don't think we need to do it in `Surface::configure()` either.